### PR TITLE
Revise in-memory database handling for modern SQLite

### PR DESCRIPTION
**Pull request by gwynne**
SQLiteKit has for a long time depended on the use of SQLite's shared cache with in-memory databases to support opening multiple connections to such databases. However, SQLite has also for a long time specifically recommended the total omission of the shared cache, a recommendation which SQLiteNIO now follows as of vapor/sqlite-nio#51. Therefore, SQLiteKit now uses uniquely-named temporary files to emulate the feature when in-memory databases are requested.

Additional changes in this release:
- To match Vapor and SQLiteNIO, Swift 5.7 is now required.
- Updated the README and the API documentation.

### CODE REVIEW

File: Sources/SQLiteKit/SQLiteConnectionSource.swift [hunk 1]
```diff
@@ -6,24 +7,26 @@ import NIOCore
...
     }
     
     public init(
         configuration: SQLiteConfiguration,
         threadPool: NIOThreadPool
     ) {
         self.configuration = configuration
+        switch configuration.storage {
+        case .memory(identifier: let identifier):
+            let filenameSafeIdentifer = String(identifier.map { $0 == ":" ? "_" : ($0 == "\\" ? "+" : ($0 == "/" ? "-" : $0)) })
```
Comment (MahdiBM) [inline Sources/SQLiteKit/SQLiteConnectionSource.swift:24]: emmmmmm can't say i like this ....... it could possibly be a dictionary lookup like:
```swift
let charMap: [Character: Character] = [
    ":": "_",
    "\\": "+",
    "/": "-",
]
Then `identifier.map { charMap[$0] ?? $0 }`
```
I'm also guessing Swift itself has some "character encoding stuff" for file names?

**Review by MahdiBM — APPROVED** [review]

File: Sources/SQLiteKit/SQLiteConnectionSource.swift (hunk 1 above)
Comment (gwynne) [reply Sources/SQLiteKit/SQLiteConnectionSource.swift]: I redid it, 'cause you were right 🙂.

File: Sources/SQLiteKit/SQLiteConnectionSource.swift [hunk 2]
```diff
@@ -48,3 +44,31 @@ public struct SQLiteConnectionSource: ConnectionPoolSource {
...
+
+fileprivate extension String {
+    var asSafeFilename: String {
+        #if os(Windows)
+        self.replacingOccurrences(of: ":", with: "_").replacingOccurrences(of: "\\", with: "-")
+        #else
+        self.replacingOccurrences(of: "/", with: "-")
+        #endif
+    }
+}
```
Comment (MahdiBM) [inline Sources/SQLiteKit/SQLiteConnectionSource.swift:56]: 👌

File: Sources/SQLiteKit/SQLiteDataEncoder.swift [hunk 3]
```diff
@@ -33,107 +30,54 @@ public struct SQLiteDataEncoder {
...
     private final class _Encoder: Encoder {
-        var codingPath: [CodingKey] {
-            return []
-        }
-
-        var userInfo: [CodingUserInfoKey : Any] {
-            return [:]
-        }
-
+        var codingPath: [any CodingKey] = []
```
Comment (MahdiBM) [inline Sources/SQLiteKit/SQLiteDataEncoder.swift:33]: why the change from computed to stored? 🤔

File: Sources/SQLiteKit/SQLiteRow+SQLRow.swift [hunk 4]
```diff
@@ -14,10 +14,10 @@ extension SQLiteRow: SQLRow {
     }
 
     public func contains(column: String) -> Bool {
-        return self.column(column) != nil
+        self.column(column) != nil
     }
 
-    public func decode<D>(column: String, as type: D.Type) throws -> D where D : Decodable {
+    public func decode<D: Decodable>(column: String, as: D.Type) throws -> D {
```
Comment (MahdiBM) [inline Sources/SQLiteKit/SQLiteRow+SQLRow.swift:20]: You could also do stuff like `(some Decodable).Type` but to be honest not sure if the current approach isn't better/more-readable.

File: Tests/SQLiteKitTests/SQLiteKitTests.swift [hunk 5]
```diff
@@ -5,102 +5,112 @@ import XCTest
 import SQLiteNIO
 import SQLKit
 
-class SQLiteKitTests: XCTestCase {
+final class SQLiteKitTests: XCTestCase {
```
Comment (MahdiBM) [inline Tests/SQLiteKitTests/SQLiteKitTests.swift:8]: Why final in _tests_?

**Review by MahdiBM — APPROVED** [review]

File: Sources/SQLiteKit/SQLiteConnectionSource.swift (hunk 2 above)
Comment (gwynne) [reply Sources/SQLiteKit/SQLiteConnectionSource.swift:56]: I even consulted with Saleem to end up with this 😉

File: Sources/SQLiteKit/SQLiteDataEncoder.swift (hunk 3 above)
Comment (gwynne) [reply Sources/SQLiteKit/SQLiteDataEncoder.swift:33]: 1) Because these properties shouldn't really be hardcoded this way and this is preparatory to a more substantial improvement to how this crud works.
2) As stored properties the fact that they're never used (even by Codable itself) doesn't count against the test coverage percentage 😂

File: Sources/SQLiteKit/SQLiteRow+SQLRow.swift (hunk 4 above)
Comment (gwynne) [reply Sources/SQLiteKit/SQLiteRow+SQLRow.swift:20]: Nope. The type has to be referred to by its generic name (`D`) at least twice in the signature (the parameter and the return type), as well as an additional time in the method body; `some` is only useful where you never need to refer directly to the type by generic name. The usage in the method body could be gotten around by referring to the parameter by its label, but there's no avoiding that you have to express it more than once in the signature. (Oh, and writing `some Decodable` twice would be the equivalent of `<T: Decodable, U: Decodable>`, which is not what's needed here.)

File: Tests/SQLiteKitTests/SQLiteKitTests.swift (hunk 5 above)
Comment (gwynne) [reply Tests/SQLiteKitTests/SQLiteKitTests.swift:8]: All classes should always be `final` unless they're explicitly intended for subclassing (not forgetting that subclassing has been soft-deprecated in Swift for a long time and subclassing `XCTestCase` is the only legitimate use of it I've yet run into in the server ecosystem). A `final` class can be better optimized, can be used covariantly and contravariantly when referring to `Self` (I won't pretend that I remember the exact rules involved in that one; if something about that seems incorrect, it probably is), and IMO is just cleaner.

File: Sources/SQLiteKit/SQLiteRow+SQLRow.swift (hunk 4 above)
Comment (MahdiBM) [reply Sources/SQLiteKit/SQLiteRow+SQLRow.swift:20]: ah right didn't notice the `-> D`

File: Sources/SQLiteKit/Docs.docc/Documentation.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+This package provides the "foundational" level of support for using [Fluent] with SQLite by implementing the requirements of an [SQLKit] driver. It is responsible for:
+
+- Managing the underlying SQLite library ([SQLiteNIO]),
+- Providing a two-way bridge between SQLiteNIO and SQLKit's generic data and metadata formats,
+- Presenting an interface for establishing, managing, and interacting with database connections.
+
+> Note: The FluentKit driver for SQLite is provided by the [FluentSQLiteDriver] package.
+
+[SQLKit]: https://swiftpackageindex.com/vapor/sql-kit
```
Comment (0xTim) [inline Sources/SQLiteKit/Docs.docc/Documentation.md:19]: Why are we pointing to the SPI out of interest?

**Review by 0xTim — APPROVED** [review]

File: Sources/SQLiteKit/Docs.docc/Documentation.md (hunk 6 above)
Comment (gwynne) [reply Sources/SQLiteKit/Docs.docc/Documentation.md:19]: If I remember my original reasoning correctly, it was because the SPI's front page for a given package is more convenient in several ways than GitHub's.

Comment (0xTim) [reply Sources/SQLiteKit/Docs.docc/Documentation.md:19]: Cool I have no objections to be clear
